Add CollectionsMarshal.GetValueRefOrAddDefault

Sergio0694 opened the pull request:
### Contributes to https://github.com/dotnet/runtime/issues/27062

```csharp
namespace System.Runtime.InteropServices
{
    public static class CollectionsMarshal
    {
        public static ref TValue? GetValueRefOrAddDefault<TKey, TValue>(Dictionary<TKey, TValue> dictionary, TKey key, out bool exists)
            where TKey : notnull;
    }
}
```

> **NOTE:** I've marked the returned `TValue` as an uncostrained nullable value, because if `TValue` is a reference type the returned reference will point to `null`. Value types are not affected by this, ie. there is no `Nullable<T>` involved here at all.

Sergio0694 committed:
Add CollectionsMarshal.GetValueRefOrAddDefault API

Sergio0694 committed:
Add special path for dictionary resize after insertion

Sergio0694 committed:
Add CollectionsMarshal.GetValueRefOrAddDefault tests

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs:
```diff
@@ -681,6 +681,178 @@ private bool TryInsert(TKey key, TValue value, InsertionBehavior behavior)
             return true;
         }
 
+        internal ref TValue? GetValueRefOrAddDefault(TKey key, out bool exists)
```
Ideally, this would not add any new code to Dictionary to make this niche method pay-for-play.

Sergio0694 replied:
I was hoping for `Dictionary` to already have an equivalent method to just expose like it was done for `GetValueRefOrAddRef`, but unfortunately there doesn't seem to be one and I'm not really seeing a way to implement this API without either adding some code (which I did so far) or at least making some changes to `TryInsert` (which is the most similar one) and using that. The reason I didn't do that though was that that would've likely led to some micro-benchmark regressions in other scenarios where the additional logic to return a ref wasn't needed. Not really sure how to proceed here, feedbacks are more than welcome 😄

One random idea I had to move the logic out of the `Dictionary` type was to refactor the code to be like:
```csharp
class Dictionary
{
    internal class Impl
    {
        public internal ref TValue? GetValueRefOrAddDefault(TKey key, out bool exists);
    }
}
```

So that `Impl.GetValueRefOrAddDefault` could access private members of the dictionary but without the method adding overhead to the dictionary type itself. Though this approach is adding another type, so it might not be desireable either. Another idea was to move the whole method logic to `CollectionsMarshal`, but that'd have required to make all the necessary dictionary members internal (and not private), so that seemed at the very least quite a bit ugly 😕

Thoughts?

stephentoub replied:
> without either adding some code (which I did so far) or at least making some changes to TryInsert (which is the most similar one) and using that. The reason I didn't do that though was that that would've likely led to some micro-benchmark regressions in other scenarios where the additional logic to return a ref wasn't needed.

Now you know why @benaadams didn't add this ;-)

Sergio0694 replied:
It was a trap 🤣

Sergio0694 commented:
@jkotas Could you clarify what the overhead of this new method would be, and whether it would be a blocker for this PR? This would be a very neat feature to have, and the API itself has been approved by API review too - if it's just the current implementation that's an issue would any of the proposed approaches help? As in, moving the code to a diferent spot, etc.

Thanks! 🙂

jkotas commented:
It makes every Dictionary instantiation more expensive, in particular with AOT.

Sergio0694 commented:
I see. If the issue is specifically with the overhead of dictionary instantiations, would moving the code to a separate type (eg. a nested type within dictionary) solve that? That would only ever be loaded when the method is actually used, right? 🤔

jkotas commented:
> would moving the code to a separate type (eg. a nested type within dictionary) solve that

Yes, it would solve that.

Sergio0694 committed:
Move GetValueRefOrAddDefault impl to separate type

This avoids the additional overhead when loading Dictionary<TKey, TValue> instances, especially in AOT scenarios, and it makes the new API pay for play.

Sergio0694 commented:
Test failures should be unrelated, and the changes in https://github.com/dotnet/runtime/pull/54611/commits/99523ba7ee43fb2c0f40f111387f837e668880f1 addresses Jan's concerns on additional overhead being added to `Dictionary<TKey, TValue>` instantiations (moved all the new code to a separate type). Should be good to review now 😄

GrabYourPitchforks reviewed (commented):
@Sergio0694 Thanks for this! Can you also add a unit test for the randomized hash logic? I think the appropriate place to add this would be in the file https://github.com/dotnet/runtime/blob/main/src/libraries/System.Collections/tests/Generic/Dictionary/HashCollisionScenarios/OutOfBoundsRegression.cs.

There's no need to modify `RunDictionaryTest` (line 93 of that file) to run a `CollectionsMarshal` test for every possible combination; that's probably overkill. Instead, what you can do is immediately _before_ line 93, insert a manual call that looks something like:

```cs
RunCollectionTestCommon(
    () => new Dictionary<string, object>(StringComparison.Ordinal),
    (dictionary, key) => CollectionsMarshal.GetRef(dictionary, key, out _) = null,
    (dictionary, key) => dictionary.ContainsKey(key),
    dictionary => dictionary.Comparer,
    expectedInternalComparerTypeBeforeCollisionThreshold: nonRandomizedOrdinalComparerType,
    expectedPublicComparerBeforeCollisionThreshold: StringComparer.Ordinal,
    expectedInternalComparerTypeAfterCollisionThreshold: randomizedOrdinalComparerType);
```

In a nutshell, this utilizes the existing test infrastructure to generate a whole bunch of collisions, but the delegate you provide will force it to use your new `CollectionsMarshal` method instead of `Add`. This will get regression test coverage for string collision scenarios in this method.

Sergio0694 commented on src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:
```diff
@@ -29,5 +29,15 @@ public static Span<T> AsSpan<T>(List<T>? list)
+        /// <param name="dictionary">The dictionary to get the ref to <typeparamref name="TValue"/> from.</param>
+        /// <param name="key">The key used for lookup.</param>
+        /// <param name="exists">Whether or not a new entry for the given key was added to the dictionary.</param>
+        /// <remarks>Items should not be added or removed from the <see cref="Dictionary{TKey, TValue}"/> while the ref <typeparamref name="TValue"/> is in use.</remarks>
```
Whoops, fixed in https://github.com/dotnet/runtime/pull/54611/commits/ab97762a21634f0d7d863ca6dc656bd12d583cb2, thanks! 😄

Sergio0694 committed:
Add unit test for randomized hash logic

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs:
```diff
@@ -681,6 +681,187 @@ private bool TryInsert(TKey key, TValue value, InsertionBehavior behavior)
             return true;
         }
 
+        /// <summary>
```
Since this is essentially duplicating the logic of `Dictionary.TryInsert`, could you add a comment in both methods advising maintainers to keep them in sync?

Sergio0694 replied:
Done in 5209a5bedff52a33454da401854502d6b967501a 🙂

eiriktsarpalis commented on src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/CollectionsMarshalTests.cs:
```diff
@@ -299,6 +299,200 @@ public void GetValueRefOrNullRefLinkBreaksOnResize()
+        {
+            // This test is the same as the one for GetValueRefOrNullRef, but it uses
+            // GetValueRefOrAddDefault instead, and also checks for incorrect additions.
+            // The two APIs should behave the same when values are already existing.
```
```suggestion
            // The two APIs should behave the same when values already exist.
```

Sergio0694 replied:
Fixed as part of 5209a5bedff52a33454da401854502d6b967501a 👍

eiriktsarpalis commented on src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/CollectionsMarshalTests.cs:
```diff
@@ -299,6 +299,200 @@ public void GetValueRefOrNullRefLinkBreaksOnResize()
+
+            Assert.False(exists);
+            Assert.Equal(3, dict.Count);
+            Assert.False(Unsafe.IsNullRef(ref entry3Ref));
```
Shouldn't there also be an assertion here validating that `entry3Ref` equals `default(Struct)`?

Sergio0694 replied:
Right, yeah it was present in the ref type case below but not here. Fixed in 21ab729485561a6c292a90ebc5a1c28182d3f64a, thanks! 😄

eiriktsarpalis reviewed (approved):
Other than minor feedback, LGTM. Thanks!

Sergio0694 committed:
Apply review feedbacks about internal comments

Sergio0694 committed:
Add missing assertion in unit tests

GrabYourPitchforks commented:
A reminder for the future - try to avoid rebase & force-push after reviews have started. Makes it a little more complicated for us since we can't rely on the GitHub web UI and instead we need to generate incremental diffs manually.

Sergio0694 commented:
Oh, I'm sorry, I didn't realize that. Will definitely avoid doing that again in the future! 😅

GrabYourPitchforks reviewed (approved):
Thanks so much! :D
